jira_integration_example/utilities/jira_integration.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2020 Google, LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module to handle interactions with a Jira server.
+
+This module defines a function to interact with Jira server when a
+monitoring notification occurs. In addition, it defines error
+classes to be used by the function.
+"""
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+class Error(Exception):
+    """Base class for all errors raised in this module."""
+
+
+class NotificationParseError(Error):
+    """Exception raised for errors in a monitoring notification format."""
+
+
+class UnknownIncidentStateError(Error):
+    """Exception raised for errors in an invalid incident state value."""
+
+
+def update_jira_based_on_monitoring_notification(jira_client, jira_project,
+                                                 notification):
+    """Updates a Jira server based off the data in a monitoring notification.
+
+    If the monitoring notification is about an open incident, a new issue (of
+    type bug) is created on the jira server that the jira client is connected
+    to under the specified jira project.
+
+    Args:
+        jira_client: A JIRA object that acts as a client which allows
+            interaction with a specific Jira server. It is used to create
+            the new Jira issue.
+        jira_project: The key or id of the Jira project under which to create
+            the Jira issue.
+        notification: The dictionary containing the notification data.
+
+    Raises:
+        UnknownIncidentStateError: If the incident state is not open or closed.
+        NotificationParseError: If notification is missing required dict key.
+        JIRAError: If error occurs when using the jira client
+    """
+
+    try:
+        incident_data = notification['incident']
+        incident_state = incident_data['state']
+        incident_condition_name = incident_data['condition_name']
+        incident_resource_name = incident_data['resource_name']
+        incident_summary = incident_data['summary']
+        incident_url = incident_data['url']
+    except KeyError:
+        raise NotificationParseError("Notification is missing required dict key")
```

Comment:
I know it'll be a little tedious to check these all separately, but I think you should. It makes the error message more useful, because then you can return the exact key that was missing.

Reply:
I just added this in, but instead of using multiple separate try / except blocks, I appended the value of the KeyError (which is the key that caused the error) to the error message to indicate the exact key that was missing.